Only use dtterm's extension where supported

jdebp opened the pull request:
`CSI` `8` `;` _h_ `;` _w_ `t` is an extension pioneered by dtterm, that not all terminals understand.  nvim is emitting it unconditionally, regardless of terminal type.  This is wrong, and has unfortunate side effects on some terminals.

When width and height are something such as (for example) 100 columns and 50 rows, this ends up emitting `CSI` `8` `;` `5` `0` `;` `1` `0` `0` `t`.  This can potentially be understood, by a terminal that does not understand the dtterm extension, as a regular DECSLPP sequence specifying a 100 _row_ screen size.  One ends up with an 100 by 100 screen, rather than the 100 by 50 one that was actually wanted.

So only emit the dtterm extension when nvim detects terminal types for terminals that are known to implement this extension, including `xterm*`, konsole, `teraterm*`, and `rxvt*`.

jdebp committed:
Fix up "linux*" terminfo entries, too.

The terminfo entry for linux only advertises 8 colours, but nvim
tries to make it display 16 colours anyway, resulting in erroneous
SGR control sequences for colours 8 and above.
The Linux kernel terminal emulator itself has actually understood
the 256-colour control sequences since version 4.8 and the
16-colour control sequences since version 4.9.
Thus we apply the same terminfo fixup as we apply for *xterm* and
*256*, to emit the 16-colour and 256-colour control sequences even
if terminfo's setaf and setab do not advertise them.

jdebp committed:
Merge branch 'master' of https://github.com/neovim/neovim

jdebp committed:
Fix copy+paste error in new test.

jdebp committed:
Merge branch 'master' of https://github.com/neovim/neovim into DECSLPP-fix

jdebp committed:
Merge branch 'master' of https://github.com/neovim/neovim

jdebp committed:
Only use dtterm's extension where supported.

This limits the use of dtterm's extension to DECSLPP to only those
terminal types where it is known to be supported.
Because it can be potentially understood as genuine DECSLPP
sequence, setting the number of lines to a number larger than 25,
which of course can cause confusion (especially if it is the width
parameter that results in this) only use it on terminals that are
known to support the dtterm extension.

jdebp committed:
Only use dtterm's extension where supported.

This limits the use of dtterm's extension to DECSLPP to only those
terminal types where it is known to be supported.
Because it can be potentially understood as genuine DECSLPP
sequence, setting the number of lines to a number larger than 25,
which of course can cause confusion (especially if it is the width
parameter that results in this) only use it on terminals that are
known to support the dtterm extension.

jdebp committed:
Only use dtterm's extension where supported.

This limits the use of dtterm's extension to DECSLPP to only those
terminal types where it is known to be supported.
Because it can be potentially understood as genuine DECSLPP
sequence, setting the number of lines to a number larger than 25,
which of course can cause confusion (especially if it is the width
parameter that results in this) only use it on terminals that are
known to support the dtterm extension.

jdebp committed:
Only use dtterm's extension where supported.

This limits the use of dtterm's extension to DECSLPP to only those
terminal types where it is known to be supported.
Because it can be potentially understood as genuine DECSLPP
sequence, setting the number of lines to a number larger than 25,
which of course can cause confusion (especially if it is the width
parameter that results in this) only use it on terminals that are
known to support the dtterm extension.

jdebp committed:
Only use dtterm's extension where supported.

This limits the use of dtterm's extension to DECSLPP to only those
terminal types where it is known to be supported.
Because it can be potentially understood as genuine DECSLPP
sequence, setting the number of lines to a number larger than 25,
which of course can cause confusion (especially if it is the width
parameter that results in this) only use it on terminals that are
known to support the dtterm extension.

jdebp committed:
Reduced the size of a comment.

jdebp committed:
rxvt (Unicode) also understands dtterm's extension.

Also moved the commentary out of a block, to in-line.

jdebp committed:
rxvt (Unicode) also understands dtterm's extension.

Also moved the commentary out of a block, to in-line.

jdebp committed:
rxvt (Unicode) also understands dtterm's extension.

Also moved the commentary out of a block, to in-line.

jdebp committed:
rxvt (Unicode) also understands dtterm's extension.

Also moved the commentary out of a block, to in-line.

jdebp committed:
Merge branch 'linux-26colour'

jdebp committed:
Merge remote-tracking branch 'upstream/master'

Conflicts:
	src/nvim/tui/tui.c

jdebp committed:
Merge remote-tracking branch 'origin/DECSLPP-fix' into DECSLPP-fix

Conflicts:
	src/nvim/tui/tui.c

brcolow commented on src/nvim/tui/tui.c:
```diff
@@ -974,14 +993,14 @@ static void fix_terminfo(TUIData *data)
     unibi_set_if_empty(ut, unibi_enter_italics_mode, "\x1b[3m");
     unibi_set_if_empty(ut, unibi_to_status_line, "\x1b]2");
-  } else if (STARTS_WITH(term, "xterm")) {
+  } else if (data->term == kTermXTerm) {
```
Nice reuse of the enum here

brcolow commented on src/nvim/tui/tui.c:
```diff
@@ -974,14 +993,14 @@ static void fix_terminfo(TUIData *data)
   }
 
-  if (STARTS_WITH(term, "xterm") || data->term == kTermRxvt) {
+  if (data->term == kTermXTerm || data->term == kTermRxvt) {
```
And here as well 👍

brcolow commented on src/nvim/tui/tui.c:
```diff
@@ -423,9 +426,16 @@ static void tui_resize(UI *ui, int width, int height)
-    char r[16];  // enough for 9999x9999
-    snprintf(r, sizeof(r), "\x1b[8;%d;%dt", height, width);
-    out(ui, r, strlen(r));
+    // Only send this control sequence extension to terminal types that we know understand it.
```
[The linter](https://travis-ci.org/neovim/neovim/jobs/228865508#L2992) complains about line length here and a couple other lines to line 434...if it is possible to shorten anything while retaining the meaning that would be swell - no big deal though.

jdebp replied:
There you go.

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -954,6 +964,15 @@ static TermType detect_term(const char *term, const char *colorterm)
+  }
+  if (STARTS_WITH(term, "teraterm")) {
+    return kTermTeraTerm;
+  }
```
Unless you have plans to use this granularity in a future patch, let's consider all of these to be xterm.

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -974,14 +993,14 @@ static void fix_terminfo(TUIData *data)
 
-  if (STARTS_WITH(term, "xterm") || data->term == kTermRxvt) {
+  if (data->term == kTermXTerm || data->term == kTermRxvt) {
     const char *normal = unibi_get_str(ut, unibi_cursor_normal);
```
This won't get entered for kTermDTTerm/kTermTeraTerm. Is that intentional? What is the effect on those terminals?

jdebp replied:
Unchanged from what it was.  It wouldn't have been entered for `dtterm` or `teraterm*` terminal types before, either.

leonerd commented:
I would imagine also any terminals using `CSI t` in its more traditional capacity as `DECSLRM` would also get confused by this sequence.

jdebp committed:
Shorten some line lengths.

jdebp committed:
Shorten some line lengths.

jdebp committed:
Shorten some line lengths.

jdebp committed:
Shorten some line lengths.

jdebp commented:
DECSLRM is `CSI` _l_ `;` _r_ `s ` .  That has its own overlaps, in particular with SCOSC; but they are not this one.

jdebp committed:
Change TUI resize to use an extended terminal capability.

... rather than hardwiring the string and testing the terminal
type every time the screen is re-sized.

jdebp committed:
Change TUI resize to use an extended terminal capability.

... rather than hardwiring the string and testing the terminal
type every time the screen is re-sized.

jdebp committed:
Change TUI resize to use an extended terminal capability.

... rather than hardwiring the string and testing the terminal
type every time the screen is re-sized.

jdebp committed:
Change TUI resize to use an extended terminal capability.

... rather than hardwiring the string and testing the terminal
type every time the screen is re-sized.

jdebp commented:
An improved approach is to have an additional pretend `resize_screen` terminfo string capability that gets added into the real terminfo set for (only) these terminal types; rather than checking the terminal type again and again.

jdebp committed:
Merge branch 'master' of https://github.com/neovim/neovim

jdebp committed:
Merge branch 'DECSLPP-fix' of ssh://github.com/jdebp/neovim into DECSLPP-fix

jdebp committed:
Accommodate type changes from 489d10c57c68183c042fd582e6c8eadbb0efc5a7

jdebp committed:
Accommodate type changes from 489d10c57c68183c042fd582e6c8eadbb0efc5a7

justinmk commented:
@leonerd Are you in favor of this change? Or was there another approach you can suggest?

Commit:
Reset the scroll region when resizing.

DECSLPP is explicitly documented as not affecting the scroll region.
The dtterm extension is not as well documented, but it is safer than not to assume that it operates similarly.

This also eliminates a pointlessly repeated test from tui_scroll().
It additionally uses a non-parameterized DECSTBM sequence when attempting to reset back to whole-screen scrolling.

Commit:
Reset the scroll region when resizing.

DECSLPP is explicitly documented as not affecting the scroll region.
The dtterm extension is not as well documented, but it is safer than not to assume that it operates similarly.

This also eliminates a pointlessly repeated test from tui_scroll().
It additionally uses a non-parameterized DECSTBM sequence when attempting to reset back to whole-screen scrolling.

jdebp committed:
Merge branch 'master' of https://github.com/neovim/neovim

jdebp committed:
Merge branch 'DECSLPP-fix' of ssh://github.com/jdebp/neovim into DECSLPP-fix
